# **Iris Dataset - Exploratory Data Analysis**

### Objective
The objective of this analysis is to explore the structure, distribution, and relationships within the Iris dataset, and derive meaningful insights using statistical and visual techniques.

# **Importing Libraries**

In this step, we import the essential Python libraries required for data manipulation, numerical computations, and visualization.

In [102]:
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Insight

- `pathlib.Path` is used to handle file paths in an operating system-independent manner, ensuring portability of the code.
- `pandas` is used for data manipulation and handling tabular data.
- `numpy` is used for numerical operations.
- `matplotlib` and `seaborn` are used for data visualization.

These libraries form the core toolkit for performing exploratory data analysis efficiently.

# **Loading the Dataset**

In this step, the dataset is loaded from a local CSV file into a pandas DataFrame for further analysis. The use of a structured file path ensures better project organization and reproducibility.

In [91]:
# Define path to dataset 
DATA_DIR = Path.cwd().parent / "data" / "raw" / "Iris.csv"

In [92]:
# Load dataset
df = pd.read_csv(DATA_DIR)

In [93]:
# Display first few rows
df.head()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa


### Insight

- The dataset has been successfully loaded into a pandas DataFrame using a relative file path, ensuring portability across different environments.
- Displaying the first few rows using `head()` provides an initial view of the dataset’s structure and feature layout.
- This step serves as a preliminary validation to confirm that the data has been correctly loaded and is ready for further exploration.

# **Dataset Overview**

In this step, we examine the structure of the dataset, including data types, non-null values, and overall composition of features.

In [94]:
df.shape

(150, 6)

In [95]:
df.columns

Index(['Id', 'SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm',
       'Species'],
      dtype='str')

In [96]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    str    
dtypes: float64(4), int64(1), str(1)
memory usage: 7.2 KB


### Insight (Dataset Structure)

- The dataset contains 150 observations and 5 columns.
- It consists of four numerical features (`SepalLengthCm`, `SepalWidthCm`, `PetalLengthCm`, `PetalWidthCm`) and one categorical feature (`Species`).
- All columns have 150 non-null values, indicating that there are no missing values in the dataset.
- The dataset appears to be clean and well-structured at a structural level.

In [97]:
df.describe()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,75.500000,5.843333,3.054000,3.758667,1.198667
std,43.445368,0.828066,0.433594,1.764420,0.763161
min,1.000000,4.300000,2.000000,1.000000,0.100000
25%,38.250000,5.100000,2.800000,1.600000,0.300000
50%,75.500000,5.800000,3.000000,4.350000,1.300000
75%,112.750000,6.400000,3.300000,5.100000,1.800000
max,150.000000,7.900000,4.400000,6.900000,2.500000


### Insight (Numerical Features Overview)

- The numerical features exhibit different value ranges, with petal-related features generally having wider variation compared to sepal features.
- The mean and median values for most features are relatively close, suggesting no strong skewness at a high level.
- The spread of the data, as indicated by standard deviation, appears moderate across features.
- No obvious anomalies or extreme values are observed from this high-level statistical summary.

In [98]:
df.describe(include='object')

C:\Users\Admin\AppData\Local\Temp\ipykernel_3896\87514550.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


,Species
count,150
unique,3
top,Iris-setosa
freq,50


### Insight (Categorical Feature Overview)

- The dataset contains one categorical feature (`Species`) with multiple unique classes.
- The frequency distribution indicates that one of the classes appears most frequently, though detailed class balance analysis will be performed in later stages.

# **Data Cleaning**

This step focuses on assessing and improving data quality by identifying missing values, duplicate entries, and non-contributory features. Appropriate measures are applied to ensure the dataset is reliable and analytically sound.

In [99]:
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

### Insight (Missing Values)

- The dataset contains no missing values across any of the features.
- This confirms data completeness and eliminates the need for imputation or removal of records.

In [100]:
df.duplicated().sum()

np.int64(0)

### Insight (Duplicate Records)

- No duplicate entries are present in the dataset.
- This ensures data consistency and avoids bias that could arise from repeated observations.

In [101]:
df.columns
df = df.drop(columns=['Id'])
df.head()

,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


### Insight (Feature Relevance)

- The `Id` column, if present, serves only as a unique identifier and does not contribute to the analysis.
- It has been removed to ensure that only meaningful features are retained for further exploration and modeling.

### Insight (Cleaning Scope)

- No missing values or duplicate records were identified in the dataset; therefore, no imputation or record removal was required.
- Advanced preprocessing techniques such as normalization, scaling, or binning were not applied at this stage, as the objective is exploratory analysis rather than model preparation.
- The dataset is already clean and structured, making it directly suitable for further exploratory analysis.